Sources/NIO/BSDSocketAPI.swift

```diff
@@ -284,3 +461,533 @@ extension NIOBSDSocket.Option {
+    internal static func getpeername(socket s: NIOBSDSocket.Handle,
+                                     address name: UnsafeMutablePointer<sockaddr>,
+                                     address_len namelen: UnsafeMutablePointer<socklen_t>)
+            throws {
+        #if os(Windows)
+            if WinSDK.getpeername(s, name, namelen) == SOCKET_ERROR {
+                throw IOError(winsock: WSAGetLastError(), reason: "getpeername")
+            }
+        #else
+            _ = try syscall(blocking: false) {
+              cGetpeername(s, name, namelen)
+            }
+        #endif
+    }
+
+    @inline(never)
+    internal static func getsockname(socket s: NIOBSDSocket.Handle,
+                                     address name: UnsafeMutablePointer<sockaddr>,
+                                     address_len namelen: UnsafeMutablePointer<socklen_t>)
+            throws {
+        #if os(Windows)
+            if WinSDK.getsockname(s, name, namelen) == SOCKET_ERROR {
+                throw IOError(winsock: WSAGetLastError(), reason: "getsockname")
+            }
+        #else
+            _ = try syscall(blocking: false) {
+                cGetsockname(s, name, namelen)
+            }
+        #endif
+    }
+
+    @inline(never)
+    internal static func getsockopt(socket: NIOBSDSocket.Handle,
+                                    level: NIOBSDSocket.OptionLevel,
+                                    option_name optname: NIOBSDSocket.Option,
+                                    option_value optval: UnsafeMutableRawPointer,
+                                    option_len optlen: UnsafeMutablePointer<socklen_t>)
+            throws {
+        #if os(Windows)
+            if WinSDK.getsockopt(socket, level.rawValue, optname.rawValue,
+                                 optval.assumingMemoryBound(to: CChar.self),
+                                 optlen) == SOCKET_ERROR {
+                throw IOError(winsock: WSAGetLastError(), reason: "getsockopt")
+            }
+        #else
+            _ = try syscall(blocking: false) {
+                cGetsockopt(socket, level.rawValue, optname.rawValue, optval, optlen)
+            }
+        #endif
+    }
+
+    @inline(never)
+    internal static func listen(socket s: NIOBSDSocket.Handle, backlog: CInt)
+            throws {
+        #if os(Windows)
+            if WinSDK.listen(s, backlog) == SOCKET_ERROR {
+              throw IOError(winsock: WSAGetLastError(), reason: "listen")
+            }
+        #else
+              _ = try syscall(blocking: false) {
+                  cListen(s, backlog)
+              }
+        #endif
+    }
+
+    @inline(never)
+    internal static func recv(socket s: NIOBSDSocket.Handle,
+                              buffer buf: UnsafeMutableRawPointer,
+                              length len: size_t) throws -> IOResult<size_t> {
+        #if os(Windows)
+            let iResult: CInt =
+                WinSDK.recv(s, buf.assumingMemoryBound(to: CChar.self),
```

Comment:
This has the same pointer concerns above: given the multiple incidences of this patten, it may be better to shim them through a helper.